File: 0087-scramble-string/0087-scramble-string.py
class Solution:
    def isScramble(self, ss1: str, ss2: str) -> bool:

        sta = {} # 记忆化存储

        def check(s1=ss1, s2=ss2):
            if len(s1) != len(s2):
                return False
            if len(s1) <= 1:
                return s1 == s2
            
            if (s1, s2) in sta:
                return sta[s1, s2]

            # from left to right
            mem = {}
            for i in range(len(s1)):
                if s1[i] not in mem:
                    mem[s1[i]] = 1
                else:
                    mem[s1[i]] += 1
                    if mem[s1[i]] == 0:
                        mem.pop(s1[i])
                
                if s2[i] not in mem:
                    mem[s2[i]] = -1
                else:
                    mem[s2[i]] -= 1
                    if mem[s2[i]] == 0:
                        mem.pop(s2[i])

                if len(mem) == 0:
                    if i != len(s1)-1:
                        if check(s1[:i+1], s2[:i+1]) and check(s1[i+1:], s2[i+1:]):
                            sta[s1, s2] = True
                            return True

            # from right to left
            mem = {}
            for i in range(len(s1)):
                if s1[i] not in mem:
                    mem[s1[i]] = 1
                else:
                    mem[s1[i]] += 1
                    if mem[s1[i]] == 0:
                        mem.pop(s1[i])
                    
                if s2[-i-1] not in mem:
                    mem[s2[-i-1]] = -1
                else:
                    mem[s2[-i-1]] -= 1
                    if mem[s2[-i-1]] == 0:
                        mem.pop(s2[-i-1])
                    
                if len(mem) == 0:
                    if i != len(s1)-1:
                        if check(s1[:i+1], s2[-i-1:]) and check(s1[i+1:], s2[:-i-1]):
                            sta[s1, s2] = True
                            return True
            sta[s1, s2] = False
            return False
            
        return check()

# 作者：Nrgeup
# 链接：https://leetcode.cn/problems/scramble-string/solutions/725200/ji-yi-hua-sou-suo-by-nrgeup-z0kv/
# 来源：力扣（LeetCode）
# 著作权归作者所有。商业转载请联系作者获得授权，非商业转载请注明出处。

File: 0087-scramble-string/0087-scramble-string.cpp
#include <iostream>
#include <unordered_map>
#include <string>
#include <vector>

using namespace std;

class Solution {
public:
    unordered_map<string, bool> sta; // 记忆化存储

    bool check(char* s1, char* s2, int len) {
        if (len <= 1) {
            return *s1 == *s2;
        }

        string key(s1, len);
        key += string(s2, len);

        if (sta.find(key) != sta.end()) {
            return sta[key];
        }

        // from left to right
        vector<int> mem(256, 0); // 使用一个数组代替unordered_map
        for (int i = 0; i < len; ++i) {
            mem[s1[i]]++;
            mem[s2[i]]--;

            bool is_valid = true;
            for (int j = 0; j < 256; ++j) {
                if (mem[j] != 0) {
                    is_valid = false;
                    break;
                }
            }

            if (is_valid) {
                if (i != len - 1) {
                    if (check(s1, s2, i + 1) && check(s1 + i + 1, s2 + i + 1, len - i - 1)) {
                        sta[key] = true;
                        return true;
                    }
                }
            }
        }

        // from right to left
        fill(mem.begin(), mem.end(), 0); // 清空计数器
        for (int i = 0; i < len; ++i) {
            mem[s1[i]]++;
            mem[s2[len - i - 1]]--;

            bool is_valid = true;
            for (int j = 0; j < 256; ++j) {
                if (mem[j] != 0) {
                    is_valid = false;
                    break;
                }
            }

            if (is_valid) {
                if (i != len - 1) {
                    if (check(s1, s2 + len - i - 1, i + 1) && check(s1 + i + 1, s2, len - i - 1)) {
                        sta[key] = true;
                        return true;
                    }
                }
            }
        }

        sta[key] = false;
        return false;
    }

    bool isScramble(string ss1, string ss2) {
        if (ss1.size() != ss2.size()) {
            return false;
        }
        
        return check(&ss1[0], &ss2[0], ss1.size());
    }
};
